nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix

```diff
@@ -0,0 +1,37 @@
+{ config, lib, pkgs }: let
+  mountOptions = { options = ["ro" "nosuid" "nodev" "bind"]; };
+  mounts = [
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-cuda-mps-control";
+      containerPath = "/usr/bin/nvidia-cuda-mps-control"; }
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-cuda-mps-server";
+      containerPath = "/usr/bin/nvidia-cuda-mps-server"; }
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-debugdump";
+      containerPath = "/usr/bin/nvidia-debugdump"; }
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-powerd";
+      containerPath = "/usr/bin/nvidia-powerd"; }
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-smi";
+      containerPath = "/usr/bin/nvidia-smi"; }
+    { hostPath = "${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk";
+      containerPath = "/usr/bin/nvidia-ctk"; }
+    { hostPath = "${pkgs.glibc}/lib";
+      containerPath = "${pkgs.glibc}/lib"; }
+  ];
+  jqAddMountExpression = ".containerEdits.mounts[.containerEdits.mounts | length] |= . +";
+  mountsToJq = lib.concatMap
+    (mount:
+      ["${pkgs.jq}/bin/jq '${jqAddMountExpression} ${builtins.toJSON (mount // mountOptions)}'"])
+    mounts;
+in ''
+#! ${pkgs.runtimeShell}
+
+function cdiGenerate {
+  ${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk cdi generate \
+    --format json \
+    --ldconfig-path ${pkgs.glibc.bin}/bin/ldconfig \
+    --library-search-path ${config.hardware.nvidia.package}/lib \
+    --nvidia-ctk-path ${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk
+}
+
+cdiGenerate | \
+  ${lib.concatStringsSep " | " mountsToJq} > $RUNTIME_DIRECTORY/nvidia-container-toolkit.json
```

Comment:
Sorry for the ping, I thought I posted a question here about `RUNTIME_DIRECTORY` and `/run/cdi` instead of `/var/run/cdi` but I cannot find it and I wonder if I only imagined posting it...

So what was the resolution about `/run` vs `/var/run`, and do we not need to provide xtra configuration for CDI to scan `/run/cdi` in addition to `/etc/cdi`?

Reply:
I think you are referring to https://github.com/NixOS/nixpkgs/pull/284507#pullrequestreview-1861131726.

I would have preferred to be strict and write to /var/run (as the spec mentions), but if the symlink /var/run -> /run is something we can assume, then I don't see any issue with it.

Reply:
> assume symlink

- [ ] We don't have to assume, we can e.g. implement a simple nixos test that puts something simple (mount a dummy path) in `/run/cdi` and verifies the configuration does have effect in containers (TODO for next PRs)

Reply:
The spec mentions explicitly `/var/run`, and the reference implementation that will probably be used by all go implementors [has it hardcoded.](https://github.com/cncf-tags/container-device-interface/blob/7b540ad63013eafbb6779fd06e3235653528d4ae/pkg/cdi/spec-dirs.go#L26-L31). As long as NixOS keeps `/var/run` a symlink to `/run` it will be fine.

Not sure why we would want to try to make CDI look somewhere else if the spec says otherwise, and the reference implementation does as well. I mean, we could, but is it worth the effort?

Reply:
I dont' see the spec saying anything about the path having to be `/var/run` specifically. The README does the search paths can be configured at least for containerd: https://github.com/cncf-tags/container-device-interface/?tab=readme-ov-file#containerd-configuration.

AFAIU the `/var/run` only exists for compatibility with legacy projects so it's weird CDI should refer to it at all...